massa-execution-worker/src/controller.rs

```diff
@@ -87,21 +95,34 @@ pub struct ExecutionControllerImpl {
 }
 
 impl ExecutionController for ExecutionControllerImpl {
-    /// called to signal changes on the current blockclique, also listing newly finalized blocks
+    /// Updates blockclique status by signaling newly finalized blocks and the latest blockclique.
     ///
-    /// # arguments
-    /// * `finalized_blocks`: list of newly finalized blocks to be appended to the input finalized blocks. Each Storage owns the block and its ops/endorsements/parents.
-    /// * `blockclique`: new blockclique, replaces the current one in the input. Each Storage owns the block and its ops/endorsements/parents.
+    /// # Arguments
+    /// * `finalized_blocks`: newly finalized blocks indexed by slot.
+    /// * `blockclique`: new blockclique (if changed). Indexed by slot.
+    /// * `block_storage`: storage instances for new blocks. Each one owns refs to the block and its ops/endorsements/parents.
     fn update_blockclique_status(
         &self,
-        finalized_blocks: HashMap<Slot, (BlockId, Storage)>,
-        new_blockclique: HashMap<Slot, (BlockId, Storage)>,
+        finalized_blocks: HashMap<Slot, BlockId>,
+        new_blockclique: Option<HashMap<Slot, BlockId>>,
+        block_storage: PreHashMap<BlockId, Storage>,
     ) {
-        // update input data
+        // lock input data
         let mut input_data = self.input_data.1.lock();
-        input_data.new_blockclique = Some(new_blockclique); // replace blockclique
-        input_data.finalized_blocks.extend(finalized_blocks); // append finalized blocks
-        self.input_data.0.notify_one(); // wake up VM loop
+
+        // extend block info
+        input_data.block_storage.extend(block_storage);
+
+        // extend finalized blocks
+        input_data.finalized_blocks.extend(finalized_blocks);
+
+        // update blockclique
+        if new_blockclique.is_some() {
+            input_data.new_blockclique = new_blockclique;
```

Comment:
I don't understand in the code where do we skip the process of the `new_blockclique` field if he is the same as before. Can you point me to this code @damip ?

Reply:
In consensus, we set new_blockclique to None if it's the same as before: https://github.com/massalabs/massa/blob/c0865fe4d79bb0f0e63c26332608b023a7697ebf/massa-consensus-worker/src/consensus_worker.rs#L608-L611

Reply:
Yes but then if it's None in consensus it will not change the blockclique here but the field `new_blockclique` still old the previous state ? And if yes what prevent it to re-compute it ?

Reply:
Exactly. We notify Execution only if there is a new blockclique. And when Execution takes the execution input data, it will always get the latest blockclique (or None if there is no new one since the last take). That way if Execution is slow, there might be multiple blockclique changes in-between takes and it won't have to process all blockcliques, just the latest one.

For final blocks it's different because they all need to be executed (new final blocks don't cancel previous ones). So we accumulate the final blocks in the execution input data so that when Execution takes them, it gets their whole accumulation to process them all.

There is also another optimization in the input data: the block_storage. Before this PR, whenever a block was finalized, we would send a full copy of its storage to Execution. Also, when the blockclique changed, we used to send a full copy of all the storages of all the blocks of the new blockclique. That was a lot of storage clones/drops. Now only send block_storage from consensus to execution for blocks that weren't sent to Execution before.